walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -25,59 +25,67 @@
         return Arrays.asList();
     }
     
+    Collection<String> allAttendees = new ArrayList();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
     // List of TimeRanges that will not work for meeting attendees
-    Collection<TimeRange> blockedTimes = new ArrayList();
+    Collection<TimeRange> blockedTimes = createListOfBlockedTimes(events, request.getAttendees());
     // List of TimeRanges that will not work for meeting attendees + optional attendees
-    Collection<TimeRange> optionalBlockedTimes = new ArrayList();
-    
-    createListOfBlockedTimes(events, request, blockedTimes);
-    createListOfBlockedTimes(events, request, optionalBlockedTimes);
+    Collection<TimeRange> optionalBlockedTimes = createListOfBlockedTimes(events, allAttendees);
 
+    
+    // List of TimeRanges that will work for meeting attendees + optional attendees
+    Collection<TimeRange> optionalAvailableTimes = createListOfAvailableTimes(events, request, optionalBlockedTimes);
     // List of TimeRanges that will work for meeting attendees
     Collection<TimeRange> availableTimes = new ArrayList();
-    // List of TimeRanges that will work for meeting attendees + optional attendees
-    Collection<TimeRange> optionalAvailableTimes = new ArrayList();
-
-    createListOfAvailableTimes(events, request, blockedTimes, availableTimes);
-    createListOfAvailableTimes(events, request, optionalBlockedTimes, optionalAvailableTimes);
+    if (optionalAvailableTimes.size() == 0) {
+        availableTimes = createListOfAvailableTimes(events, request, blockedTimes);
+    }
+    
 
-    // Only return the List that works for optional attendees if there is one or more TimeRange
+    // Only return the List that works for optional attendees if there is one or more TimeRange or no attendees
     if (optionalAvailableTimes.size() > 0) {
         return optionalAvailableTimes;
+    } else if (optionalAvailableTimes.size() == 0 && (request.getAttendees()).size() == 0) {
+        return optionalAvailableTimes;
     }
     return availableTimes;
   }
 
   /**
   * Create a List of TimeRanges that will not work for the given attendees based on events that are scheduled.
   */
-  public void createListOfBlockedTimes(Collection<Event> events, MeetingRequest request, Collection<TimeRange> blockedTimes) {
+  public Collection<TimeRange> createListOfBlockedTimes(Collection<Event> events, Collection<String> totalAttendees) {
+    // List of TimeRanges that will not work for given meeting attendees
+    Collection<TimeRange> blockedTimes = new ArrayList();
     // Add the unavailable TimeRanges into the blockedTimes list
-    for(String attendee: request.getAttendees()) {
+    for(String attendee: totalAttendees) {
         for(Event event: events){
             if(event.getAttendees().contains(attendee)) {
                 blockedTimes.add(event.getWhen());
             }
         }
     }
+    return blockedTimes;
```

Comment:
consider using a stream and sets instead of collections

https://www.baeldung.com/java-stream-filter-lambda
https://www.geeksforgeeks.org/sets-intersection-function-guava-java/

i think you could do something like 
```
events.toStream()
    .filter(event -> !Sets.intersection(event.getAttendees(), totalAttendees).isEmpty())
    .map(event -> event.getWhen())
    .collect(Collectors.toList())
```

It looks like `getAttendees()` returns a set, so you can also make `totalAttendees` a set.

Reply:
keeping the original code - talked w/ Kevin about the issue.